File: javatest/src/main/java/patterns/AdapterApp.java
package patterns;

import java.io.FileNotFoundException;

public class AdapterApp {

    public static void main(String[] args) throws FileNotFoundException {
        //1-й способ через наследование
        VectorGraphicsInterface g1 = new VectorAdapterFromRaster();
        g1.drawLine();
        g1.drawSquare();
        //2-й способ через композицию
        VectorGraphicsInterface g2 = new VectorAdapterFromRaster2(new RasterGraphics());
        g2.drawLine();
        g2.drawSquare();
    }
}

//--------------------------------Adaptee (интерфейс который будем адаптировать)---------------------------------------
class RasterGraphics {
    void drawRasterLine() {
        System.out.println("Рисуем линию");
    }

    void drawRasterSquare() {
        System.out.println("Рисуем квадрат");
    }
}

//--------------------------------Adapter (интерфейс который будем использовать)---------------------------------------
interface VectorGraphicsInterface {
    void drawLine();

    void drawSquare();
}

//ConcreteAdapter 1. Реализация через наследование
class VectorAdapterFromRaster extends RasterGraphics implements VectorGraphicsInterface {
    public void drawLine() {
        drawRasterLine();
    }

    public void drawSquare() {
        drawRasterSquare();
    }
}

//ConcreteAdapter 2. Реализация через композицию
class VectorAdapterFromRaster2 implements VectorGraphicsInterface {
    RasterGraphics raster = null;//new RasterGraphics();

    public VectorAdapterFromRaster2(RasterGraphics raster) {
        this.raster = raster;
    }

    public void drawLine() {
        raster.drawRasterLine();
    }

    public void drawSquare() {
        raster.drawRasterSquare();
    }
}

File: _kotlin_/src/main/kotlin/patterns/AdapterApp.kt
package patterns

import java.io.FileNotFoundException

object AdapterApp {

    @Throws(FileNotFoundException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        //1-й способ через наследование
        val g1: VectorGraphicsInterface = VectorAdapterFromRaster()
        g1.drawLine()
        g1.drawSquare()
        //2-й способ через композицию
        val g2: VectorGraphicsInterface = VectorAdapterFromRaster2(RasterGraphics())
        g2.drawLine()
        g2.drawSquare()
    }
}

//--------------------------------Adaptee (интерфейс который будем адаптировать)---------------------------------------
open class RasterGraphics {

    fun drawRasterLine() {
        println("Рисуем линию")
    }

    fun drawRasterSquare() {
        println("Рисуем квадрат")
    }
}

//--------------------------------Adapter (интерфейс который будем использовать)---------------------------------------
interface VectorGraphicsInterface {

    fun drawLine()
    fun drawSquare()
}

//ConcreteAdapter 1. Реализация через наследование
class VectorAdapterFromRaster : RasterGraphics(), VectorGraphicsInterface {

    override fun drawLine() {
        drawRasterLine()
    }

    override fun drawSquare() {
        drawRasterSquare()
    }
}

//ConcreteAdapter 2. Реализация через композицию
class VectorAdapterFromRaster2(raster: RasterGraphics?) :
    VectorGraphicsInterface {

    var raster: RasterGraphics? = null //new RasterGraphics();

    init {
        this.raster = raster
    }

    override fun drawLine() {
        raster!!.drawRasterLine()
    }

    override fun drawSquare() {
        raster!!.drawRasterSquare()
    }
}
